File: Python/10 - Excepciones/Q10_devJPallares.py
'''
 * EJERCICIO:
 * Explora el concepto de manejo de excepciones según tu lenguaje.
 * Fuerza un error en tu código, captura el error, imprime dicho error
 * y evita que el programa se detenga de manera inesperada.
 * Prueba a dividir "10/0" o acceder a un índice no existente
 * de un listado para intentar provocar un error.
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea una función que sea capaz de procesar parámetros, pero que también
 * pueda lanzar 3 tipos diferentes de excepciones (una de ellas tiene que
 * corresponderse con un tipo de excepción creada por nosotros de manera
 * personalizada, y debe ser lanzada de manera manual) en caso de error.
 * - Captura todas las excepciones desde el lugar donde llamas a la función.
 * - Imprime el tipo de error.
 * - Imprime si no se ha producido ningún error.
 * - Imprime que la ejecución ha finalizado.
 '''

x = 10
y: list[int] = [1, 2, 3, 4]

def excepciones(x: int, y: list[int]) -> str:
    
    if not isinstance(x, int):
        raise TypeError(f"The first Parameter: x=(\"{x}\"), must be an integer")
    elif not isinstance(y, list) or not all(isinstance(i, int) for i in y):
        raise TypeError(f"The second Parameter: y=(\"{y}\"), must be contain a list of integers")
    else:
        print(f'Divividendo: {x}')
        print(f'Divisores: {y}\n{"-"*10}')
        for i in y:
            print(f'{x} / {i} = {x/i:.1f}')
    
try:
    excepciones(x, y)

except NameError as e:
    print(f'Error type: "{e}" [The variable "{e.name}" is not defined]')
except ZeroDivisionError as e:
    print(f'Error type: "{e}" [You can\'t divide by zero]')
except TypeError as e:
    print(f'Error type: "{e}" [The data type of the any parameter: \n(x: "{type(x)}", y: "{type(y)}" or list contained values in "{y}" isn\'t integer]')
else:
    print('Completed without exceptions')
finally:
    print('Exceptions testing finished')

File: Java/src/Topic10_Exceptions/Q10_devJPallares.java

// * DIFICULTAD EXTRA (opcional):
// * Crea una función que sea capaz de procesar parámetros, pero que también
// * pueda lanzar 3 tipos diferentes de excepciones (una de ellas tiene que
// * corresponderse con un tipo de excepción creada por nosotros de manera
// * personalizada, y debe ser lanzada de manera manual) en caso de error.
// * - Captura todas las excepciones desde el lugar donde llamas a la función.
// * - Imprime el tipo de error.
// * - Imprime si no se ha producido ningún error.
// * - Imprime que la ejecución ha finalizado.

package Topic10_Exceptions;

import java.awt.*;
import java.util.InputMismatchException;
import java.util.Scanner;

public class Q10_devJPallares {

    static Integer capt_dividendo () {
        Scanner capDiv = new Scanner(System.in);

        System.out.print("Digite el dividendo: ");
        String a = capDiv.nextLine(); // Leer como texto primero

        try {
            return Integer.parseInt(a); // Intentar convertir
        } catch (NumberFormatException e) {
            throw new InputMismatchException("The Parameter must be an integer");
        }
    }

    static void exceptions(Integer b, int[] c) {

        for (int i : c) {
            if (i == 0) {
                throw new ArithmeticException("Can't divide by zero");
            } else {
                System.out.printf("%d // %d = %d\n", b, i, b / i);
            }
        }
        System.out.println("------------------------------------------");
    }

    public static void main(String[] args) {
        int[] y = {1, 2, 3, 4};

        try {
            int x = capt_dividendo();
            System.out.println("------------------------------------------");
            System.out.printf("Dividendo: %d\n", x);
            System.out.printf("Divisores: ");
            for (int j : y) {
                System.out.printf("[%d]", j);
            }
            System.out.printf("\n------------------------------------------\n");

            exceptions(x,y);

        } catch (ArithmeticException e) {
            System.out.printf("Error: \n%s", e);
        } catch (InputMismatchException e) {
            System.out.printf("Error: \n%s", e);
        } catch (Exception e) {
            System.out.printf("Error: \n%s", e);
        } finally {
            System.out.printf("\nSe completó la ejecución del programa");
        }
    }
}
